### PR TITLE
[SPARK-27899][SQL] Make HiveMetastoreClient.getTableObjectsByName available in ExternalCatalog/SessionCatalog API

### PR DESCRIPTION
**Pull request by LantaoJin**
## What changes were proposed in this pull request?

The new Spark ThriftServer SparkGetTablesOperation implemented in https://github.com/apache/spark/pull/22794 does a catalog.getTableMetadata request for every table. This can get very slow for large schemas (~50ms per table with an external Hive metastore).
Hive ThriftServer GetTablesOperation uses HiveMetastoreClient.getTableObjectsByName to get table information in bulk, but we don't expose that through our APIs that go through Hive -> HiveClientImpl (HiveClient) -> HiveExternalCatalog (ExternalCatalog) -> SessionCatalog.

If we added and exposed getTableObjectsByName through our catalog APIs, we could resolve that performance problem in SparkGetTablesOperation.

## How was this patch tested?

Add UT

### CODE REVIEW

**Comment by LantaoJin** [comment]
Hi @wangyum @gatorsmile , could you have a chance to review?

**Comment by juliuszsompolski** [comment]
Thank you @LantaoJin for doing that!
I looked through the code and it looks good to me, but I am not really familiar with the catalog code so will refrain from reviewing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -434,6 +434,22 @@ class SessionCatalog(
...
   }
 
+  /**
+   * Retrieve the metadata of an existing permanent table/view. If no database is specified,
+   * assume the table/view is in the current database.
+   */
+  @throws[NoSuchDatabaseException]
+  def getTablesByName(names: Seq[TableIdentifier]): Seq[CatalogTable] = {
+    if (names.nonEmpty) {
+      val db = formatDatabaseName(names.head.database.getOrElse(getCurrentDatabase))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: What if names from different databases? for example:
```scala
Seq("db1.table1", "db2.table1")
```

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 2]
```diff
@@ -156,6 +156,7 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
 
     withJdbcStatement("table1", "table2") { statement =>
       Seq(
+        "DROP VIEW IF EXISTS view1",
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala]: Is this change necessary?

**Comment by wangyum** [comment]
@LantaoJin Could you post some benchmark result?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala [hunk 3]
```diff
@@ -128,6 +128,8 @@ trait ExternalCatalog {
 
   def getTable(db: String, table: String): CatalogTable
 
+  def getTablesByName(db: String, tables: Seq[String]): Seq[CatalogTable]
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: When adding a new API like this, we need to add various test cases in ExternalCatalogSuite that is extended by HiveExternalCatalogSuite and InMemoryCatalogSuite.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -384,127 +388,144 @@ private[hive] class HiveClientImpl(
...
-        val sortColumnNames = if (allAscendingSorted) {
-          sortColumnOrders.map(_.getCol)
-        } else {
-          Seq.empty
-        }
-        Option(BucketSpec(h.getNumBuckets, h.getBucketCols.asScala, sortColumnNames))
+    getRawTableOption(dbName, tableName).map(convertHiveTableToCatalogTable)
+  }
+
+  private def convertHiveTableToCatalogTable(h: HiveTable): CatalogTable = {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Could you submit a refactor PR first? This can reduce the code changes made by this PR.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 5]
```diff
@@ -237,6 +237,10 @@ class VersionsSuite extends SparkFunSuite with Logging {
       assert(client.getTableOption("default", "src").isDefined)
     }
 
+    test(s"$version: getTablesByName") {
+      assert(client.getTablesByName("default", Seq("src")).nonEmpty)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: This is just a special case. We need more to ensure all the hive metastore versions can correctly support it . Also negative cases are needed. :-)

**Comment by gatorsmile** [comment]
Thank you for your fast work! Will review the details after you address the above comments.

**Comment by gatorsmile** [comment]
ok to test

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 2 above)
Comment (LantaoJin) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala]: This view was not cleaned up when the UT failed in my environment. So adding this could stable the test case I think.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 1 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: I have considered this case. Could this be happened in the reality use cases? But good catch, it should be handled in code. How about throws an exception?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 4 above)
Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I would like to do. Does it needs to create a new issue or not?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 5 above)
Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Sorry, what means? Hasn’t it been covered from 0.12 to 3.1?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 4 above)
Comment (gatorsmile) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Nope. You can reuse this issue.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 5 above)
Comment (gatorsmile) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: this only covers a single table. How about multiple tables? or mixed with non-existent tables? with illegal table names? or an empty seq? and so on. :-)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 4 above)
Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: See #24803

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala (hunk 3 above)
Comment (LantaoJin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 6]
```diff
@@ -434,6 +434,31 @@ class SessionCatalog(
...
+   * For example, if none of the requested tables could be retrieved, an empty list is returned.
+   * There is no guarantee of ordering of the returned tables.
+   */
+  @throws[NoSuchDatabaseException]
+  def getTablesByName(names: Seq[TableIdentifier]): Seq[CatalogTable] = {
+    if (names.nonEmpty) {
+      val dbs = names.map(_.database.getOrElse(getCurrentDatabase))
+      if (dbs.distinct.size != 1) {
+        throw new AnalysisException(
+          s"Only the tables/views belong to one same database can be retrieved."
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Let us list the input name list in the error message

**Comment by LantaoJin** [comment]
After rebased from master, the commit history contains many unrelated commits. I have to force push from a new branch to clean them.

**Comment by LantaoJin** [comment]
Looks good now

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 7]
```diff
@@ -434,6 +434,34 @@ class SessionCatalog(
...
+   * For example, if none of the requested tables could be retrieved, an empty list is returned.
+   * There is no guarantee of ordering of the returned tables.
+   */
+  @throws[NoSuchDatabaseException]
+  def getTablesByName(names: Seq[TableIdentifier]): Seq[CatalogTable] = {
+    if (names.nonEmpty) {
+      val dbs = names.map(_.database.getOrElse(getCurrentDatabase))
+      if (dbs.distinct.size != 1) {
+        val tables = names.map(name => formatTableName(name.table))
+        dbs.zip(tables).map { case (d, t) => QualifiedTableName(d, t)}
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 8]
```diff
@@ -434,6 +434,34 @@ class SessionCatalog(
...
+        throw new AnalysisException(
+          s"Only the tables/views belong to one same database can be retrieved. Querying " +
+          s"tables/views are ${dbs.zip(tables).map { case (d, t) => QualifiedTableName(d, t)}}"
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: This should be replaced by a variable name.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 9]
```diff
@@ -277,6 +277,20 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
     }
   }
 
+  test("get tables by name") {
+    assert(newBasicCatalog().getTablesByName("db2", Seq("tbl1", "tbl2"))
+      .map(_.identifier.table) == Seq("tbl1", "tbl2"))
+  }
+
+  test("get tables by name when some tables do not exists") {
+    assert(newBasicCatalog().getTablesByName("db2", Seq("tbl1", "tblnotexist"))
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: See `validNameFormat`.  Add a test case when the seq of table names contains the invalid name.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala]: Weird. If the view is not dropped, why the tables do not exist in your environment?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClient.scala [hunk 10]
```diff
@@ -84,6 +84,8 @@ private[hive] trait HiveClient {
   /** Returns the metadata for the specified table or None if it doesn't exist. */
   def getTableOption(dbName: String, tableName: String): Option[CatalogTable]
 
+  def getTablesByName(dbName: String, tableNames: Seq[String]): Seq[CatalogTable]
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClient.scala]: Add a comment to describe the function?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 11]
```diff
@@ -237,6 +237,25 @@ class VersionsSuite extends SparkFunSuite with Logging {
       assert(client.getTableOption("default", "src").isDefined)
     }
 
+    test(s"$version: getTablesByName") {
+      assert(client.getTablesByName("default", Seq("src")).head
+        == client.getTableOption("default", "src").get)
+    }
+
+    test(s"$version: getTablesByName when multiple tables") {
+      assert(client.getTablesByName("default", Seq("src", "temporary"))
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Also try the invalid names here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 12]
```diff
@@ -1091,6 +1114,38 @@ private[hive] object HiveClientImpl {
...
+        if ((b.length == 1) && (b(0) < 10)) { // ^A, ^B, ^C, ^D, \t
+          parameters.put(serdeConstants.SERIALIZATION_FORMAT, Integer.toString(b(0)))
+        }
+      }
+      // Use LazySimpleSerDe for MetadataTypedColumnsetSerDe.
+      // NOTE: LazySimpleSerDe does not support tables with a single column of col
+      // of type "array<string>". This happens when the table is created using
+      // an earlier version of Hive.
+      if (classOf[MetadataTypedColumnsetSerDe].getName
+            == tTable.getSd.getSerdeInfo.getSerializationLib
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: 4-space indent 

== needs to be moved to the line 1139

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 13]
```diff
@@ -1091,6 +1114,38 @@ private[hive] object HiveClientImpl {
...
+   * of [[org.apache.hadoop.hive.ql.metadata.Hive]] to do some extra fixes for non-views.
+   * Methods of extracting multiple [[HiveTable]] like `getRawTablesByName()`
+   * should invoke this before return.
+   */
+  def extraFixesForNonView(tTable: MetaStoreApiTable): MetaStoreApiTable = {
+    // For non-views, we need to do some extra fixes
+    if (!(HiveTableType.VIRTUAL_VIEW.toString == tTable.getTableType)) {
+      // Fix the non-printable chars
+      val parameters: JMap[String, String] = tTable.getSd.getParameters
+      val sf: String = parameters.get(serdeConstants.SERIALIZATION_FORMAT)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: https://github.com/apache/hive/blame/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java#L1342 

Do we need to check null here?

**Review by gatorsmile — COMMENTED** [review]
Thanks for your work!

**Comment by wangyum** [comment]
I did a simple benchmark in our production environment(The **default** database has `1626` tables):
```scala
cat <<EOF > SPARK-27899.scala
def benchmark(func: () => Unit): Long = {
  val start = System.currentTimeMillis()
  for(i <- 0 until 2) { func() }
  val end = System.currentTimeMillis()
  end - start
}

def default(): Unit = {
  val list = new java.util.ArrayList[Array[AnyRef]]()
  val catalog = spark.sessionState.catalog
  catalog.listTables("default").foreach { tableIdentifier =>
    val catalogTable = catalog.getTableMetadata(tableIdentifier)
    val rowData = Array[AnyRef](
      "",
      catalogTable.database,
      catalogTable.identifier.table,
      catalogTable.tableType,
      catalogTable.comment.getOrElse(""))
    list.add(rowData)
  }
}

def spark_27899(): Unit = {
  val list = new java.util.ArrayList[Array[AnyRef]]()
  val catalog = spark.sessionState.catalog
  catalog.getTablesByName(catalog.listTables("default")).foreach { catalogTable =>
    val rowData = Array[AnyRef](
      "",
      catalogTable.database,
      catalogTable.identifier.table,
      catalogTable.tableType,
      catalogTable.comment.getOrElse(""))
    list.add(rowData)
  }
}

val defaultTimeToken = benchmark(() => default)
val spark27899TimeToken = benchmark(() => spark_27899)
println(s"Default time token: $defaultTimeToken")
println(s"SPARK-27899 time token: $spark27899TimeToken")
EOF
```
Benchmark result:
```
Default time token: 317983
SPARK-27899 time token: 58977
```

**Comment by LantaoJin** [comment]
https://github.com/apache/spark/pull/24774#discussion_r291036823
Seems it's my IDE problem. Cleaned and gone. I will remove this "drop view"

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClient.scala (hunk 10 above)
Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClient.scala]: added

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 9 above)
Comment (LantaoJin) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: added

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 11 above)
Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: added

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master. cc @juliuszsompolski
